sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala

```diff
@@ -252,7 +252,7 @@ trait Row extends Serializable {
    *
    * @throws ClassCastException when data type does not match.
    */
-  def getDate(i: Int): java.sql.Date = apply(i).asInstanceOf[java.sql.Date]
+  def getDate(i: Int): java.sql.Date = DateUtils.toJavaDate(getInt(i))
```

Comment:
one thing - you probably want to do the conversion when we create the row, like what we do for other types, instead of doing the conversion when it is accessed. otherwise apply(i: Int) will return date.